core/Pos/Binary/Core/Address.hs

```diff
@@ -44,59 +41,41 @@ backwards-compatible with and so we could special-case those types of
-sizeNPutAddressIncomplete :: (Size Address, Address -> Poke ())
-sizeNPutAddressIncomplete = convertToSizeNPut toBi
-  where
-    toBi :: Address -> PokeWithSize ()
-    toBi = \case
-        -- It's important that we use 'putWithTag' for all branches.
-        -- See the note above.
-        PubKeyAddress keyHash attrs ->
-            putWithTag 0 $
-                putS keyHash <>
-                putAttributesS addrToList attrs
-        ScriptAddress scrHash ->
-            putWithTag 1 $ putS scrHash
-        RedeemAddress keyHash ->
-            putWithTag 2 $ putS keyHash
-        UnknownAddressType t bs ->
-            -- It's important that it's 'putBytesS' and not just 'putS'.
-            -- See the note above.
-            putWithTag t $ putBytesS bs
-
-    -- | Put tag, then length of X, then X itself
-    putWithTag :: Word8 -> PokeWithSize () -> PokeWithSize ()
-    putWithTag t x = putWord8S t <> putSmallWithLengthS x
-
-    addrToList :: AddrPkAttrs -> [(Word8, PokeWithSize ())]
-    addrToList = \case
-        AddrPkAttrs Nothing -> []
-        AddrPkAttrs (Just path) -> [(0, putS path)]
-
--- | Decode everything except for CRC32
-getAddressIncomplete :: Peek Address
-getAddressIncomplete = do
-    tag <- getWord8
-    getSmallWithLength $ \len -> case tag of
-        0 -> do
-            let mapper 0 x = Just $ get <&> \a -> x {addrPkDerivationPath = Just a}
-                mapper _ _ = Nothing
-            PubKeyAddress <$> get <*> getAttributes mapper Nothing def
-        1 -> ScriptAddress <$> get
-        2 -> RedeemAddress <$> get
-        t -> UnknownAddressType t <$> getBytes (fromIntegral len)
-
 instance CRC32 Address where
-    crc32Update seed = crc32Update seed . encodeWithS sizeNPutAddressIncomplete
+    crc32Update seed = crc32Update seed . serialize'
+
+----------------------------------------
+
+instance Bi (Attributes AddrPkAttrs) where
+    encode = encodeAttributes [(0, serialize' . addrPkDerivationPath)]
+    decode = decodeAttributes def $ \n v acc -> case n of
+        0 -> Just $ acc { addrPkDerivationPath = deserialize' v }
+        _ -> Nothing
 
 instance Bi Address where
-    sizeNPut = labelS "Address" $
-        sizeNPutAddressIncomplete <>
-        putField crc32
-    get = label "Address" $ do
-       addr <- getAddressIncomplete
-       checksum <- get
-       if checksum /= crc32 addr
-           then fail "Address has invalid checksum!"
-           else return addr
+    encode addr =
+        encodeListLen 2 <> encodeAddr
+      where
+        encodeAddr = case addr of
+            PubKeyAddress keyHash attrs ->
+                   encode (0 :: Word8)
+                <> encode (serialize' (keyHash, attrs))
+            ScriptAddress scrHash ->
+                   encode (1 :: Word8)
+                <> encode (serialize' scrHash)
+            RedeemAddress keyHash ->
+                   encode (2 :: Word8)
+                <> encode (serialize' keyHash)
```

Comment:
@gromakovsky This is part of the general pattern where you have an `Unknown` constructor. In this case you want to serialise everything as a `ByteString` blob with `serialize'` in order to be able to be resilient to schema changes. In practice, if we cannot figure out how to deserialize something, instead of failing, we can always yield an `UnknownAddress tag blob`.